bundles/org.openhab.binding.solarwatt/src/main/java/org/openhab/binding/solarwatt/internal/channel/SolarwattChannelTypeProvider.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.solarwatt.internal.channel;
+
+import java.util.Collection;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.measure.Unit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.solarwatt.internal.SolarwattBindingConstants;
+import org.openhab.binding.solarwatt.internal.domain.SolarwattChannel;
+import org.openhab.core.library.CoreItemFactory;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.type.*;
+import org.openhab.core.types.StateDescriptionFragmentBuilder;
+import org.openhab.core.types.util.UnitUtils;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link ChannelTypeProvider} that creates {@link ChannelType}s according to
+ * the requested tags. It creates one {@link ChannelType} per tag value.
+ *
+ * @author Matthias Steigenberger - Initial contribution
+ * @author Sven Carstens - Adapted to solarwatt binding
+ *
+ */
+@NonNullByDefault
+@Component(service = { ChannelTypeProvider.class, SolarwattChannelTypeProvider.class })
+public class SolarwattChannelTypeProvider implements ChannelTypeProvider {
+
+    private final Logger logger = LoggerFactory.getLogger(SolarwattChannelTypeProvider.class);
+
+    private final Map<String, ChannelType> channelMap = new ConcurrentHashMap<>();
+
+    @Override
+    public Collection<ChannelType> getChannelTypes(@Nullable Locale locale) {
+        return this.channelMap.values();
+    }
+
+    @Override
+    public @Nullable ChannelType getChannelType(ChannelTypeUID channelTypeUID, @Nullable Locale locale) {
+        return this.channelMap.values().stream().filter(channelType -> channelType.getUID().equals(channelTypeUID))
+                .findFirst().orElse(null);
+    }
+
+    /**
+     * Assert that the {@link ChannelType} matching our requirements exists.
+     *
+     * Only create once for each tagname supplied via {@link SolarwattChannel}.
+     * 
+     * @param solarwattChannel channeltype requirements
+     * @return UID of existing channeltype
+     */
+    public ChannelTypeUID assertChannelType(SolarwattChannel solarwattChannel) {
+        @Nullable
+        ChannelType existingChannel = this.channelMap.get(solarwattChannel.getChannelName());
```

Comment:
The compiler is more intelligent on local variables. Therefore, the Nullable could be removed. Please check all.

Reply:
I remove the Nullable and NotNull annotations. Code still compiles without extra warnings but IntelliJ starts nagging.